libnet/src/ioctl.rs

```diff
@@ -1269,59 +1269,53 @@ pub(crate) fn create_ip_addr_static(
         }
 
         match addr {
-            IpPrefix::V6(_) => {
+            IpNet::V6(_) => {
                 create_logical_interface_v6(sock, &mut req)?;
             }
-            IpPrefix::V4(_) => {
+            IpNet::V4(_) => {
                 create_logical_interface_v4(sock, &mut req)?;
             }
         }
 
         // assign addr
         match addr {
-            IpPrefix::V6(a) => {
+            IpNet::V6(a) => {
                 req.lifr_lifru.lifru_addr.ss_family = AF_INET6 as addr_family_t;
                 let sas =
                     &mut req.lifr_lifru.lifru_addr as *mut sockaddr_storage;
 
                 let sa6 = sas as *mut sockaddr_in6;
-                (*sa6).sin6_addr.s6_addr = a.addr.octets();
+                (*sa6).sin6_addr.s6_addr = a.addr().octets();
             }
-            IpPrefix::V4(a) => {
+            IpNet::V4(a) => {
                 req.lifr_lifru.lifru_addr.ss_family = AF_INET as addr_family_t;
                 let sas =
                     &mut req.lifr_lifru.lifru_addr as *mut sockaddr_storage;
                 let sa4 = sas as *mut sockaddr_in;
-                (*sa4).sin_addr.s_addr = Into::<u32>::into(a.addr).to_be();
+                let addr: u32 = a.addr().into();
+                (*sa4).sin_addr.s_addr = addr.to_be();
             }
         };
         ioctl_ro!(sock, sys::SIOCSLIFADDR, &req)?;
 
         // assign netmask
         match addr {
-            IpPrefix::V6(a) => {
+            IpNet::V6(a) => {
                 req.lifr_lifru.lifru_addr.ss_family = AF_INET6 as addr_family_t;
                 let sas =
                     &mut req.lifr_lifru.lifru_addr as *mut sockaddr_storage;
 
                 let sa6 = sas as *mut sockaddr_in6;
-                let mut addr: u128 = 0;
-                for i in 0..a.mask {
-                    addr |= 1 << (127 - i);
-                }
-                (*sa6).sin6_addr.s6_addr = Ipv6Addr::from(addr).octets();
+                (*sa6).sin6_addr.s6_addr = a.mask_addr().octets();
             }
-            IpPrefix::V4(a) => {
+            IpNet::V4(a) => {
                 req.lifr_lifru.lifru_addr.ss_family = AF_INET as addr_family_t;
                 let sas =
                     &mut req.lifr_lifru.lifru_addr as *mut sockaddr_storage;
 
                 let sa4 = sas as *mut sockaddr_in;
-                let mut addr: u32 = 0;
-                for i in 0..a.mask {
-                    addr |= 1 << i;
-                }
```

Comment:
this seems inverted (i.e the logic above for ipv6 seems correct)

do I have this right? How could I validate this? Thanks!

Reply:
Yep, good catch!

You can test as follows

current behavior:

```
ry@masaka:~/src/netadm-sys$ pfexec netadm create addr igb0/test 10.2.3.4/24
ERROR netadm:ioctl: ioctl @sock.as_raw_fd()=3 sys::SIOCSLIFNETMASK: Operation not supported (os error 48)
```

with this patch:

```
ry@masaka:~/src/netadm-sys$ pfexec ./target/debug/netadm create addr igb0/test 10.2.3.4/24
ry@masaka:~/src/netadm-sys$ ipadm show-addr igb0/test
ADDROBJ           TYPE     STATE        ADDR
igb0/test         static   ok           10.2.3.4/24
```

This little program is also helpful

```rust
use std::net::Ipv4Addr;

fn main() {
    let len = 4;
    let mut m: u32 = 0;
    for i in 0..len {
        m |= 1 << (31 - i);
    }
    let msk = Ipv4Addr::from(m);
    println!("{msk}");

    m = m.to_be();
    println!("{m:032b}");

    let pfx: oxnet::Ipv4Net = "1.0.0.0/4".parse().unwrap();
    println!("{}", pfx.mask_addr());
    m = pfx.mask_addr().into();
    m = m.to_be();
    println!("{m:032b}");
}
```

Reply:
I think I may be missing something. `pfexec netadm create addr igb0/test 10.2.3.4/24` works for me, and I **think** the existing logic works for any prefix length that's divisible by 8 because we get 0x00_ff_ff_ff little endian which turns into 255.255.255.0 in network byte order. But with, say, 10.2.3.4/22 that's 0x00_3f_ff_ff -> 255.255.63.0 ... which is invalid.